app/forms/ConsignmentReferences.scala

```diff
@@ -38,14 +39,41 @@ object ConsignmentReferences {
 
   val allowedReferenceAnswers: Seq[String] = Seq(Ducr, Mucr)
 
-  val mapping = Forms.mapping(
-    "reference" -> requiredRadio("consignmentReferences.reference.empty")
-      .verifying("consignmentReferences.reference.error", isContainedIn(allowedReferenceAnswers)),
-    "referenceValue" -> text()
-      .verifying("consignmentReferences.reference.value.empty", nonEmpty)
-      .verifying("consignmentReferences.reference.value.error", isEmpty or validDucrOrMucr)
-  )(ConsignmentReferences.apply)(ConsignmentReferences.unapply)
+  val mapping = Forms
+    .mapping(
+      "reference" -> requiredRadio("consignmentReferences.reference.empty")
+        .verifying("consignmentReferences.reference.error", isContainedIn(allowedReferenceAnswers)),
+      "referenceValue" -> text()
+        .verifying("consignmentReferences.reference.value.empty", nonEmpty)
+    )(ConsignmentReferences.apply)(ConsignmentReferences.unapply)
 
   def form(): Form[ConsignmentReferences] = Form(mapping)
 
 }
+
+object ConsignmentReferencesForm {
+
+  def bindFromRequest(implicit request: play.api.mvc.Request[_]) = {
+    val baseForm = ConsignmentReferences.form().bindFromRequest
+
+    if (baseForm.errors.nonEmpty) {
+      baseForm
+    } else {
+      baseForm.value
+        .map(
+          reference =>
+            if (referenceValid(reference)) { baseForm } else {
+              baseForm.withError("referenceValue", "consignmentReferences.reference.value.error")
+          }
+        )
+        .getOrElse(baseForm)
```

Comment:
A lot of logic here, I think we'll be able to make it shorter or some kind easier. Why this reference can't be validated in the mapping?

You can use verifying when you build whole object, not just for elements

Reply:
I can be validated in the mapping.  But you can't then include the key to the field that's wrong (and therefore it doesn't get highlighted)